### PR TITLE
Overhaul `-Ztreat-err-as-bug`

### PR DESCRIPTION
**Pull request by nnethercote**
It's current behaviour is surprising, in a bad way. This also makes the implementation more complex than it needs to be.

r? @oli-obk

### CODE REVIEW

**Comment by nnethercote** [comment]
Let me explain in some detail. Consider this program with a single type error.
```
fn main() { let a: u32 = true; }
```

Compile it normally and you get this:
```
error[E0308]: mismatched types
 --> a.rs:1:26
  |
1 | fn main() { let a: u32 = true; }
  |                    ---   ^^^^ expected `u32`, found `bool`
  |                    |
  |                    expected due to this
```

Ok, that's fine.

Now compile it with `-Ztreat-err-as-bug=1` and you get this:
```
error: internal compiler error: `TypeError` when attempting coercion but no error emitted
 --> a.rs:1:26
  |
1 | fn main() { let a: u32 = true; }
  |                          ^^^^

thread 'rustc' panicked at compiler/rustc_errors/src/lib.rs:1467:30:
aborting due to `-Z treat-err-as-bug=1`
```

There is a type error, and the compiler panics afterwards. But the error is labelled as an ICE, there is no error code, the error message is different, and it is missing the "expected" markers. Huh?

Now compile it with `-Ztreat-err-as-bug=2` and you get this:
```
error[E0308]: mismatched types
 --> a.rs:1:26
  |
1 | fn main() { let a: u32 = true; }
  |                    ---   ^^^^ expected `u32`, found `bool`
  |                    |
  |                    expected due to this

thread 'rustc' panicked at compiler/rustc_errors/src/lib.rs:1471:25:
aborting after 1 errors and 1 delayed bugs due to `-Z treat-err-as-bug=2`
```

The type error is back to the original: it's a normal error instead of an ICE, the error message matches, the error code is there, and the "expected" markers are there. And again it panics afterwards. This is the behaviour I would expect with `-Ztreat-err-as-bug`. What is going on?

The final line of the last example explains it: "aborting after 1 errors and 1 delayed bugs". The compiler has "delayed bugs": a kind of second-class diagnostic that only gets printed at the end if no other errors are printed. They're widely used but mostly invisible; you never see them unless there's a bug in the compiler. And yet, `-Ztreat-err-as-bug` treats delayed bugs the same as normal errors.

So now we can now understand the `-Ztreat-err-as-bug=1` output. When handling this type error, the compiler first emits a delayed bug, and then emits a normal error. It's showing the delayed bug, which has a different message, less span information, and happens to be treated as an ICE rather than a normal error followed by a panic.

I think this behaviour is very confusing and not what anybody is likely to want in practice. This PR changes `-Ztreat-err-as-bug` so it ignores delayed bugs, unless they get promoted to proper bugs and are printed.

cc @compiler-errors

**Comment by compiler-errors** [comment]
This is somewhat disappointing since it doesn't (afaict) give you the capability to intercept a delayed bug. I may add a `-Zemit-delayed-bug-as-err` to pair with this, for compiler debugging purposes.

**Comment by nnethercote** [comment]
If I wait for #119872 to merge, then I can use `-Zeagerly-emit-delayed-bugs` to undo the `storage-return.rs` test changes, which is nice.

File: tests/ui/mir/lint/storage-return.rs [hunk 1]
```diff
@@ -1,7 +1,3 @@
-// compile-flags: -Zlint-mir -Ztreat-err-as-bug
-// failure-status: 101
-// dont-check-compiler-stderr
-// error-pattern: has storage when returning
```
[the remark this answers is not in the document]
Comment (nnethercote) [reply tests/ui/mir/lint/storage-return.rs]: #119872 has merged so I was able to use `-Zeagerly-emit-delayed-bugs` and restore this test to its original form.

**Comment by nnethercote** [comment]
I updated and adjusted some of the test changes back by using the new `-Zeagerly-emit-delayed-bugs`. Should be good to go now.

**Review by compiler-errors — APPROVED** [review]
r=me when green

**Comment by nnethercote** [comment]
It's green.

@bors r=compiler-errors

**Pull request merged**